sparse/_compressed/compressed.py

```diff
@@ -137,9 +151,13 @@ def __init__(
             raise ValueError("data must be a scalar or 1-dimensional.")
 
         self.shape = shape
-        self.compressed_axes = (
-            tuple(compressed_axes) if isinstance(compressed_axes, Iterable) else None
-        )
+
+        if not isinstance(self, Compressed2d):
+            self.compressed_axes = (
+                tuple(compressed_axes)
+                if isinstance(compressed_axes, Iterable)
+                else None
+            )
```

Comment:
Why is this change necessary?

Reply:
In particular, I don't like it because it makes the superclass depend on the subclass -- That's bad coding practice, usually.

Reply:
Otherwise `self.compressed_axes` would be mutable for `Compressed2d` classes, since we'd have assignment here.

I'm open to other suggestions. Alternatives I've considered are:

* Separate `__init__` methods – my initial solution
* `setattr(csr. "compressed_axes", x)` is a no-op – but it really should be an error
* Check that the new value is correct – this is adding moving parts to an immutable property. I don't like making changes to the subclasses based purely on the implementation of super classes.
* Higher level super class. This might be adding too much indirection.

> I don't like it because it makes the superclass depend on the subclass

I agree, but think that super classes have to be designed around allowing subclassing. I think this is a perennial python problem (e.g. how easy is it to subclass an ndarray or dataframe?).

Reply:
One way to go about this is to make it immutable in this PR. Just rename it to `_compressed_axes` and also add a `@property`.

Reply:
Another is to just have it mutable for now, and make it immutable later.

Reply:
I don't like the idea of modifying GCXS to support CSR, CSC too much. Especially since I fully expect CSR, CSC code to be volatile for a bit. I also really like keeping CSR, CSC implementation as simple as possible (e.g. `compressed_axes` is constant).

How about I split the shared code into an outside function? It's a bit of indirection, but I don't think we can have a single method that works for one class.

I've also been thinking about it more, and I'm not sure `issubtype(Compressed2d, GCXS)` is always going to be a good idea. For instance, my impression is that `GCXS`, `CSF` and how `taco`'s representations for these arrays are not equivalent though can be equivalent `CSR` and `CSC`. If either of the other implementations happens, does it still make sense for `CSR`/ `CSC` to subtype `GCXS`?

Reply:
We can use composition instead of inheritance and only subclass, `SparseArray`.

Reply:
To answer your question: I don't see anyone relying on subclass behaviour too much, just the fact that it has all the operations.

Reply:
> just the fact that it has all the operations.

Yeah, this is an important part. I would like to think more about the type structure before making any big decisions on it, and would like to get this merged so operations can be defined and test/ benchmark suites can be built. I imagine experience while adding to the classes – as well as getting more eyes on it – will help clarify here.

To that end, what's the minimum fix here? To me it's:

1. Leave as is
2. Split common initialization into it's own function

I like **1**, since it's the smallest amount of code that allows both behaviors. It also has very little indirection. I'd be happy to implement **2** if that's a better solution to you. Either of these would come with an issue to think about type structure.

> I don't see anyone relying on subclass behaviour too much

I think it will come up, especially as typing continues to mature. I bet it's going to come up while writing the dispatch methods.

Reply:
I think the best fix is to leave it mutable (https://github.com/pydata/sparse/pull/442#discussion_r595011333) and move mutability to the next PR. I'll mark the issue as a release blocker if immutability is that important to you. `GCXS` wasn't released, and so we're fine changing that behavior.

As a side note we may be overthinking the immutability, `COO` is most often used, and it holds even `data` and `coords` as attributes. Hardly anyone I've come across modifies them.

Reply:
I've added a commit where it's kept as immutable, but we don't need the instance check. Just makes sure it's not set to anything other that what it should be.